q2_composition/_ancombc.py

```diff
@@ -0,0 +1,159 @@
+                         % (parameter, value))
+
+
+def _ancombc(table, metadata, formula, p_adj_method, prv_cut, lib_cut,
+             level_ordering, neg_lb, tol, max_iter, conserve, alpha):
+
+    meta = metadata.to_dataframe()
+
+    # error on IDs found in table but not in metadata
+    missing_ids = table.index.difference(meta.index).values
+
+    if not (set(table.index).issubset(set(meta.index))):
+        raise KeyError('Not all samples present within the table were found in'
+                       ' the associated metadata file. Please make sure that'
+                       ' all samples in the FeatureTable are also present in'
+                       ' the metadata.'
+                       ' Sample IDs not found in the metadata: %s'
+                       % missing_ids)
+
+    # column validation for the formula parameter
+    formula_terms = _parse_terms(formula=formula)
+    for term in formula_terms:
+        _column_validation(term, 'formula', meta)
+
+    # column & level validation for the level_ordering parameter
+    if level_ordering is not None:
+        for i in level_ordering:
+            column = i.split('::')[0]
+            level_value = i.split('::')[1]
+
+            _column_validation(column, 'level_ordering', meta)
+
+            if level_value not in pd.unique(meta[column].values):
+                raise ValueError('Value provided in `level_ordering` parameter'
+                                 ' not found in the associated column within'
+                                 ' the metadata. Please make sure each'
+                                 ' column::value pair is present within the'
+                                 ' metadata file.'
+                                 ' \n\n'
+                                 ' column::value pair with a value that was'
+                                 ' not found: "%s"' % i)
+    else:
+        level_ordering = ''
+
+    with tempfile.TemporaryDirectory() as temp_dir_name:
+        temp_dir_name = '.'
+        biom_fp = os.path.join(temp_dir_name, 'input.biom.tsv')
+        meta_fp = os.path.join(temp_dir_name, 'input.map.txt')
+
+        table.to_csv(biom_fp, sep='\t', header=True)
+        meta.to_csv(meta_fp, sep='\t', header=True)
+
+        output_loaf = DataLoafPackageDirFmt()
+
+        cmd = ['run_ancombc.R',
+               '--inp_abundances_path', biom_fp,
+               '--inp_metadata_path', meta_fp,
+               '--formula', str(formula),
+               '--p_adj_method', p_adj_method,
+               '--prv_cut', str(prv_cut),
+               '--lib_cut', str(lib_cut),
+               '--level_ordering', str(level_ordering),
+               '--neg_lb', str(neg_lb),
+               '--tol', str(tol),
+               '--max_iter', str(max_iter),
+               '--conserve', str(conserve),
+               '--alpha', str(alpha),
+               '--output_loaf', str(output_loaf)
+               ]
+
+        try:
```

Comment:
I would imagine we could actually wrap this whole thing in some framework util not jut the `run_commands` function.